clang/runtime/dpct-rt/include/dpct/device.hpp

```diff
@@ -648,22 +649,23 @@ class dev_mgr {
     check_id(id);
     return *_devs[id];
   }
+
   unsigned int current_device_id() const {
-   std::lock_guard<std::recursive_mutex> lock(m_mutex);
-   auto it=_thread2dev_map.find(get_tid());
-   if(it != _thread2dev_map.end())
-      return it->second;
-    return DEFAULT_DEVICE_ID;
+    if (_dev_stack.empty())
+      return DEFAULT_DEVICE_ID;
+    return _dev_stack.top();
   }
 
   /// Select device with a device ID.
   /// \param [in] id The id of the device which can
   /// be obtained through get_device_id(const sycl::device).
   void select_device(unsigned int id) {
-    std::lock_guard<std::recursive_mutex> lock(m_mutex);
-    check_id(id);
-    _thread2dev_map[get_tid()] = id;
+    /// Replace the top of the stack with the given device id
+    if (!_dev_stack.empty())
+      pop_device();
+    push_device(id);
```

Comment:
```suggestion
    if (_dev_stack.empty()) {
      push_device(id);
    } else {
      check_id(id);
      _dev_stack.top() = id;
    }
```

Reply:
Aren't they both are doing the same thing?

Reply:
pop and then push will do more things. Integer assignment is much simpler.